Regression(277723@main) libcpp assertions got disabled on down-levels

cdumez opened the pull request:
#### dc2c6e2bbb6ae611016071c598c2be97cbb0b936
<pre>
Regression(277723@main) libcpp assertions got disabled on down-levels
<a href="https://bugs.webkit.org/show_bug.cgi?id=273795">https://bugs.webkit.org/show_bug.cgi?id=273795</a>

Reviewed by David Kilzer.

libcpp assertions got disabled on down-levels in 277723@main because older
versions of Xcode do not support `CLANG_CXX_STANDARD_LIBRARY_HARDENING`.

Re-introduce the old way to enable the assertions (in addition to new new
one).

* Configurations/CommonBase.xcconfig:

Canonical link: <a href="https://commits.webkit.org/278441@main">https://commits.webkit.org/278441@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09129b0a23e767896c5dba5f6dc9f21a17aa904e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50557 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29853 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2866 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53816 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1247 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36104 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41218 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52656 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27501 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43531 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22328 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8935 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46881 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55405 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25655 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/759 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48626 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43684 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47677 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11071 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27780 "Built successfully") | | | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/7321 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/26648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Configurations/CommonBase.xcconfig:
```diff
@@ -59,3 +59,21 @@ WK_COMMON_WARNING_CFLAGS = -Wall -Wc99-designator -Wconditional-uninitialized -W
+WK_LIBCPP_ASSERTIONS_CFLAGS = $(WK_LIBCPP_ASSERTIONS_CFLAGS_$(WK_PLATFORM_NAME));
+WK_LIBCPP_ASSERTIONS_CFLAGS_iphoneos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17));
+WK_LIBCPP_ASSERTIONS_CFLAGS_iphonesimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17));
+WK_LIBCPP_ASSERTIONS_CFLAGS_IOS_SINCE_17 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```
```suggestion
WK_LIBCPP_ASSERTIONS_CFLAGS_iphoneos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17)$(WK_IOS_18));
WK_LIBCPP_ASSERTIONS_CFLAGS_iphonesimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17)$(WK_IOS_18));
WK_LIBCPP_ASSERTIONS_CFLAGS_IOS_SINCE_17_IOS_BEFORE_18 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```

ddkilzer commented on Configurations/CommonBase.xcconfig:
```diff
@@ -59,3 +59,21 @@ WK_COMMON_WARNING_CFLAGS = -Wall -Wc99-designator -Wconditional-uninitialized -W
+WK_LIBCPP_ASSERTIONS_CFLAGS_iphonesimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_IOS_17));
+WK_LIBCPP_ASSERTIONS_CFLAGS_IOS_SINCE_17 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
+WK_LIBCPP_ASSERTIONS_CFLAGS_maccatalyst = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACCATALYST_14));
+WK_LIBCPP_ASSERTIONS_CFLAGS_MACCATALYST_SINCE_14 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```
```suggestion
WK_LIBCPP_ASSERTIONS_CFLAGS_maccatalyst = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACCATALYST_14)$(WK_MACCATALYST_15));
WK_LIBCPP_ASSERTIONS_CFLAGS_MACCATALYST_SINCE_14_MACCATALYST_BEFORE_15 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```

ddkilzer commented on Configurations/CommonBase.xcconfig:
```diff
@@ -59,3 +59,21 @@ WK_COMMON_WARNING_CFLAGS = -Wall -Wc99-designator -Wconditional-uninitialized -W
+WK_LIBCPP_ASSERTIONS_CFLAGS_maccatalyst = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACCATALYST_14));
+WK_LIBCPP_ASSERTIONS_CFLAGS_MACCATALYST_SINCE_14 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
+WK_LIBCPP_ASSERTIONS_CFLAGS_macosx = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACOS_1400));
+WK_LIBCPP_ASSERTIONS_CFLAGS_MACOS_SINCE_1400 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```
```suggestion
WK_LIBCPP_ASSERTIONS_CFLAGS_macosx = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_MACOS_1400)$(WK_MACOS_1500));
WK_LIBCPP_ASSERTIONS_CFLAGS_MACOS_SINCE_1400_MACOS_BEFORE_1500 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```

ddkilzer commented on Configurations/CommonBase.xcconfig:
```diff
@@ -59,3 +59,21 @@ WK_COMMON_WARNING_CFLAGS = -Wall -Wc99-designator -Wconditional-uninitialized -W
+WK_LIBCPP_ASSERTIONS_CFLAGS_MACOS_SINCE_1400 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
+WK_LIBCPP_ASSERTIONS_CFLAGS_appletvos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_TVOS_17));
+WK_LIBCPP_ASSERTIONS_CFLAGS_appletvsimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_TVOS_17));
+WK_LIBCPP_ASSERTIONS_CFLAGS_TVOS_SINCE_17 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```
```suggestion
WK_LIBCPP_ASSERTIONS_CFLAGS_appletvos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_TVOS_17)$(WK_TVOS_18));
WK_LIBCPP_ASSERTIONS_CFLAGS_appletvsimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_TVOS_17)$(WK_TVOS_18));
WK_LIBCPP_ASSERTIONS_CFLAGS_TVOS_SINCE_17_TVOS_BEFORE_18 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```

ddkilzer commented on Configurations/CommonBase.xcconfig:
```diff
@@ -59,3 +59,21 @@ WK_COMMON_WARNING_CFLAGS = -Wall -Wc99-designator -Wconditional-uninitialized -W
+WK_LIBCPP_ASSERTIONS_CFLAGS_TVOS_SINCE_17 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
+WK_LIBCPP_ASSERTIONS_CFLAGS_watchos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_WATCHOS_10));
+WK_LIBCPP_ASSERTIONS_CFLAGS_watchsimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_WATCHOS_10));
+WK_LIBCPP_ASSERTIONS_CFLAGS_WATCHOS_SINCE_10 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```
```suggestion
WK_LIBCPP_ASSERTIONS_CFLAGS_watchos = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_WATCHOS_10)$(WK_WATCHOS_11));
WK_LIBCPP_ASSERTIONS_CFLAGS_watchsimulator = $(WK_LIBCPP_ASSERTIONS_CFLAGS$(WK_WATCHOS_10)$(WK_WATCHOS_11));
WK_LIBCPP_ASSERTIONS_CFLAGS_WATCHOS_SINCE_10_WATCHOS_BEFORE_11 = -D_LIBCPP_ENABLE_ASSERTIONS=1;
```

ddkilzer reviewed (approved):
I think we should limit the macros to just the current major version of each OS (iOS 17, macOS 14, tvOS 17, watchOS 10).

r=me with the attached changes.

cdumez committed:
Regression(277723@main) libcpp assertions got disabled on down-levels
https://bugs.webkit.org/show_bug.cgi?id=273795

Reviewed by David Kilzer.

libcpp assertions got disabled on down-levels in 277723@main because older
versions of Xcode do not support `CLANG_CXX_STANDARD_LIBRARY_HARDENING`.

Re-introduce the old way to enable the assertions (in addition to new new
one).

* Configurations/CommonBase.xcconfig:

Canonical link: https://commits.webkit.org/278441@main
